### PR TITLE
Porting RNN from Linen to NNX

### PR DESCRIPTION
**Pull request by zinccat**
# Porting RNN from Linen to NNX

Fixes # (4259), https://github.com/google/flax/issues/4259

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [x] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/issues/4259)
- [x] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: flax/nnx/nn/recurrent.py [hunk 1]
```diff
@@ -0,0 +1,920 @@
...
+A = TypeVar("A")
+Array = jax.Array
+Output = Any
+
+
+class RNNCellBase(Module):
+    """RNN cell base class."""
+
+    def initialize_carry(
+        self, rngs: rnglib.Rngs | None, input_shape: tuple[int, ...]
```
Comment (cgarciae) [inline flax/nnx/nn/recurrent.py:54]: Maybe we could have `input_shape` first and make `rngs` optional.
```suggestion
        self, input_shape: tuple[int, ...], rngs: rnglib.Rngs | None = None
```

File: flax/nnx/nn/recurrent.py [hunk 2]
```diff
@@ -0,0 +1,920 @@
...
+          rng: random number generator passed to the init_fn.
+          input_shape: a tuple providing the shape of the input to the cell.
+        Returns:
+          An initialized carry for the given RNN cell.
+        """
+        batch_dims = input_shape[:-1]
+        if rngs is None:
+            rngs = self.rngs
+        mem_shape = batch_dims + (self.hidden_features,)
+        c = self.carry_init(rngs.params(), mem_shape, self.param_dtype)
```
Comment (cgarciae) [inline flax/nnx/nn/recurrent.py:209]: Maybe we could use the `carry` stream as these are not technically trainable params?
```suggestion
        c = self.carry_init(rngs.carry(), mem_shape, self.param_dtype)
```

Comment (cgarciae) [reply flax/nnx/nn/recurrent.py:209]: Same for all other cells.

File: flax/nnx/nn/recurrent.py [hunk 3]
```diff
@@ -0,0 +1,920 @@
...
+    """
+
+    def __init__(
+        self,
+        cell: RNNCellBase,
+        time_major: bool = False,
+        return_carry: bool = False,
+        reverse: bool = False,
+        keep_order: bool = False,
+        unroll: int = 1,
```
Comment (cgarciae) [inline flax/nnx/nn/recurrent.py:608]: Maybe we could accepts `rngs` to get a default that can be optionally override during `__call__`?
```suggestion
        unroll: int = 1,
        rngs: rnglib.Rngs | None = None,
```

File: flax/nnx/nn/recurrent.py [hunk 4]
```diff
@@ -0,0 +1,920 @@
...
+            return carry, y
+
+        scan = nnx.scan(
+            scan_fn,
+            in_axes=(Carry, time_axis),
+            out_axes=(Carry, (0, time_axis)) if slice_carry else (Carry, time_axis),
+            unroll=self.unroll,
+        )
+
+        scan_output = scan(carry, inputs)
```
Comment (cgarciae) [inline flax/nnx/nn/recurrent.py:688]: Currently `self` is being passed as a capture, we need to pass `cell` as an explicit input.
```suggestion
        def scan_fn(cell: RNNCellBase, carry: Carry, x: Array) -> tuple[Carry, Array]:
            carry, y = cell(carry, x)
            if slice_carry:
                return carry, (carry, y)
            return carry, y

        state_axes = nnx.StateAxes({...: Carry})
        scan = nnx.scan(
            scan_fn,
            in_axes=(state_axes, Carry, time_axis),
            out_axes=(Carry, (0, time_axis)) if slice_carry else (Carry, time_axis),
            unroll=self.unroll,
        )

        scan_output = scan(self.cell, carry, inputs)
```

File: flax/nnx/nn/recurrent.py [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+    return_carry: bool = False
+
+    def __init__(
+        self,
+        forward_rnn: RNNBase,
+        backward_rnn: RNNBase,
+        *,
+        merge_fn: Callable[[Array, Array], Array] = _concatenate,
+        time_major: bool = False,
+        return_carry: bool = False,
```
Comment (cgarciae) [inline flax/nnx/nn/recurrent.py:851]: Accept `rngs` here as well.

**Comment by cgarciae** [comment]
Thanks @zinccat for doing this, this is amazing! 
Left a few comments.

**Comment by zinccat** [comment]
Thanks for the review! Will fix it soon

File: flax/nnx/nn/recurrent.py (hunk 3 above)
Comment (zinccat) [reply flax/nnx/nn/recurrent.py:608]: changed and set default to rngs(0)

**Comment by IvyZX** [comment]
Thank you for making the change! You probably need to rebase to the current head to resolve the Read the Docs build error.

**Comment by zinccat** [comment]
hi, any updates on this?

**Comment by IvyZX** [comment]
Sorry about the delay! Merging them rn

**Review by IvyZX — APPROVED** [review]

**Comment by zinccat** [comment]
thanks!

**Pull request merged**